### PR TITLE
[#142305] New Product Access List CSV export

### PR DESCRIPTION
**Pull request by wangminghao**
Creates a new Report that is a CSV export for data about Product Access lists and send it to the email address supplied by the user.  Importing Product Access Lists will come in a follow up PR.

### CODE REVIEW

File: app/models/reports/instrument_access_list_export_raw.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    def build_row(product_user)
+      user = product_user.user
+      [
+        user.username,
+        user.full_name,
+        suspended_or_expired(user),
+        user.email,
+        product_user.product_access_group.try(:name),
+        format_usa_datetime(product_user.approved_at),
```
Comment (rdunlop) [inline app/models/reports/instrument_access_list_export_raw.rb:65]: I'm confused why the header uses `:date_added`, but the row uses `approved_at`.

File: app/views/product_users/index.html.haml [hunk 2]
```diff
@@ -20,6 +20,12 @@
   %p.notice= text("none")
 - else
   - if @product.has_product_access_groups?
+    = form_tag facility_instrument_users_path(format: :csv), method: :get, class: "search_form" do
+      = hidden_field_tag :email, current_user.email, disabled: true
+      = hidden_field_tag :format, params[:format], disabled: true
```
Comment (rdunlop) [inline app/views/product_users/index.html.haml:25]: this should be unnecessary, since you have `format: :csv`. is it necessary?

File: app/views/product_users/index.html.haml [hunk 3]
```diff
@@ -20,6 +20,12 @@
   %p.notice= text("none")
 - else
   - if @product.has_product_access_groups?
+    = form_tag facility_instrument_users_path(format: :csv), method: :get, class: "search_form" do
+      = hidden_field_tag :email, current_user.email, disabled: true
```
Comment (rdunlop) [inline app/views/product_users/index.html.haml:24]: Having a hidden field like this is a smell. Can we instead use `current_user.email` in the receiving controller method?

File: app/views/product_users/index.html.haml [hunk 4]
```diff
@@ -20,6 +20,12 @@
   %p.notice= text("none")
 - else
   - if @product.has_product_access_groups?
+    = form_tag facility_instrument_users_path(format: :csv), method: :get, class: "search_form" do
```
Comment (rdunlop) [inline app/views/product_users/index.html.haml:23]: I'm confused...where's the action button for this form? Appears to be related to the link_to...hmm..do we need the form? can it be a button only?

**Review by rdunlop — CHANGES_REQUESTED** [review]

File: app/models/reports/instrument_access_list_export_raw.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+require "csv"
+
+module Reports
+
+  class InstrumentAccessListExportRaw
```
Comment (giladshanan) [inline app/models/reports/instrument_access_list_export_raw.rb:7]: Instruments aren't the only product that can have an access list, and Export Raw is an existing feature.  How about:
```suggestion
  class ProductAccessListCsvReport
```

File: app/models/reports/instrument_access_list_export_raw.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    def text_content
+      text(".body")
+    end
+
+    def has_attachment?
+      true
+    end
+
+    def translation_scope
+      "reports.instruments.product_users"
```
Comment (giladshanan) [inline app/models/reports/instrument_access_list_export_raw.rb:41]:
```suggestion
      "reports.product_users"
```

File: app/models/reports/instrument_access_list_export_raw.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      CSV.generate do |csv|
+        csv << headers
+        @product_users.each do |product_user|
+          csv << build_row(product_user)
+        end
+      end
+    end
+
+    def filename
+      "instrument_access_list.csv"
```
Comment (giladshanan) [inline app/models/reports/instrument_access_list_export_raw.rb:25]: How about we include the product name here:
```suggestion
      "#{product.name}_access_list.csv"
```
You could pass in the product on initialization.

File: app/models/reports/instrument_access_list_export_raw.rb [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        user.email,
+        product_user.product_access_group.try(:name),
+        format_usa_datetime(product_user.approved_at),
+      ]
+    end
+
+    def suspended_or_expired(user)
+      return "" if user.active?
+      user.suspended? ? "Suspended" : "Expired"
+    end
```
Comment (giladshanan) [inline app/models/reports/instrument_access_list_export_raw.rb:72]: I think I'd prefer calling this column "User Status":
```suggestion
    def user_status(user)
      if user.active?
        "Active"
      elsif user.suspended?
         "Suspended"
      else
         "Expired"
       end
    end
```

File: app/views/product_users/index.html.haml (hunk 4 above)
Comment (giladshanan) [inline app/views/product_users/index.html.haml:23]: This needs to be pulled up to line 22, out of the conditional, so the link appears regardless of the presence of product access groups

File: app/views/product_users/index.html.haml (hunk 3 above)
Comment (giladshanan) [reply app/views/product_users/index.html.haml:24]: It looks like `CsvEmailAction` defaults to `current_user.email`, and we don't need to prompt the user for an email address, so I think we can drop the hidden fields.

**Review by giladshanan — CHANGES_REQUESTED** [review]
Can you add a spec or two?  Something simple like `global_user_roles_report_spec.rb` just to cover the happy path basics.

Let me know if you want to talk any of this through in real time.

File: app/models/reports/instrument_access_list_export_raw.rb (hunk 1 above)
Comment (giladshanan) [reply app/models/reports/instrument_access_list_export_raw.rb:65]: This is probably the result of an existing confusing practice of using `ProductUser.human_attribute_name(:date_added)` to get "Date added".  There is no `date_added` attribute.

This is only used in a couple other places, let's clean it up.  I think we should use i18n instead.

File: app/views/product_users/index.html.haml [hunk 9]
```diff
@@ -20,6 +20,12 @@
...
+      = hidden_field_tag :email, current_user.email, disabled: true
+      = hidden_field_tag :format, params[:format], disabled: true
+
+    = link_to t("reports.instruments.product_users.export"), url_for(format: :csv), class: "js--exportSearchResults pull-right", data: { form: ".search_form" }
```
Comment (giladshanan) [inline app/views/product_users/index.html.haml:27]: Let's keep this simple and just send the email to the current user instead of prompting.

File: app/views/product_users/index.html.haml (hunk 4 above)
Comment (wangminghao) [reply app/views/product_users/index.html.haml:23]: Oh I thought maybe we wouldn't want to display a link that would produce an empty file - but yeah I can see that it's good for consistency with how the rest of the site functions.

File: app/views/product_users/index.html.haml (hunk 9 above)
Comment (wangminghao) [reply app/views/product_users/index.html.haml:27]: I thought this was the consistent way of doing things for the other Export to CSV links that I saw? Also, as an admin might I want to send the file to someone else?

File: app/views/product_users/index.html.haml (hunk 4 above)
Comment (wangminghao) [reply app/views/product_users/index.html.haml:23]: This is the mechanism that other Export to CSV links in the app are using. They have a form with hidden fields that a javascript snippet populates through a pop-up input box and that's how we are able to pass the input email on to the backend through an xhr call. I was duplicating this functionality (though I agree it is not ideal)

File: app/views/product_users/index.html.haml (hunk 9 above)
Comment (giladshanan) [reply app/views/product_users/index.html.haml:27]: Good points, let's keep the prompt.

File: app/views/product_users/index.html.haml (hunk 4 above)
Comment (giladshanan) [reply app/views/product_users/index.html.haml:23]: If we have `@product_users` then we should have rows in the CSV export.  Putting the link on line 22 shouldn't create an empty file.  Is it possible you're mixing up `ProductAccessGroup` and `ProductUser`?

File: app/views/product_users/index.html.haml [hunk 10]
```diff
@@ -13,12 +13,20 @@
 
 %p= text("explanation", product_type: @product.model_name.human.downcase)
 
-%p= link_to text("add"), [:new, current_facility, @product, :user], class: "btn-add"
+%p
+  = link_to text("add"), [:new, current_facility, @product, :user], class: "btn-add"
```
Comment (giladshanan) [inline app/views/product_users/index.html.haml:17]: Is this change intended?  Seems like the `%p` doesn't need it's own line.

**Review by giladshanan — APPROVED** [review]
LGTM!  One small non-blocking question.

File: app/views/product_users/index.html.haml (hunk 10 above)
Comment (giladshanan) [inline app/views/product_users/index.html.haml:17]:
```suggestion
%p = link_to text("add"), [:new, current_facility, @product, :user], class: "btn-add"
```

File: app/views/product_users/index.html.haml [hunk 11]
```diff
@@ -13,12 +13,20 @@
...
+
```
Comment (giladshanan) [inline app/views/product_users/index.html.haml:18]:
```suggestion
%p= link_to text("add"), [:new, current_facility, @product, :user], class: "btn-add"
```
